Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs

```diff
@@ -0,0 +1,293 @@
+    /// instance of this class, but <see cref="TeardownFunctions"/> must not be called concurrently
+    /// with any other calls into this class. The intention is to enable tests to spin up multiple
+    /// functions simultaneously. This is useful because function startup can be the dominant
+    /// factor in test execution time for integration tests.
+    /// </para>
+    /// </remarks>
+    public sealed class FunctionsController
+    {
+        private const long StartupTimeout = 60;
+
+        private readonly List<FunctionOutputBufferHandler> output = new List<FunctionOutputBufferHandler>();
+        private readonly object sync = new object();
+
+        /// <summary>
+        /// Start a functions instance.
+        /// </summary>
+        /// <param name="testDirectory">The directory associated with the current test context,
+        /// usually available from your testing framework's context object(s).</param>
+        /// <param name="path">The location of the functions project.</param>
+        /// <param name="port">The port on which to start the functions instance.</param>
+        /// <param name="runtime">The runtime version, defaults to netcoreapp2.1.</param>
+        /// <param name="provider">The functions provider. Defaults to csharp.</param>
+        /// <param name="configuration">A <see cref="FunctionConfiguration"/> instance, for conveying
+        /// configuration values via environment variables to the function host process.</param>
+        /// <returns>A task that completes once the function instance has started.</returns>
+        public async Task StartFunctionsInstance(
+            string testDirectory,
+            string path,
+            int port,
+            string runtime = "netcoreapp2.1",
+            string provider = "csharp",
+            FunctionConfiguration? configuration = null)
+        {
+            Console.WriteLine($"Starting a function instance for project {path} on port {port}");
+            Console.WriteLine("\tStarting process");
+
+            FunctionOutputBufferHandler bufferHandler = StartFunctionHostProcess(
+                port,
+                provider,
+                await GetToolPath(),
+                GetWorkingDirectory(testDirectory, path, runtime),
+                configuration);
+
+            lock (this.sync)
+            {
+                this.output.Add(bufferHandler);
+            }
+
+            Console.WriteLine("\tProcess started; waiting for initialisation to complete");
+
+            await Task.WhenAny(
+                bufferHandler.JobHostStarted,
+                bufferHandler.ExitCode,
+                Task.Delay(TimeSpan.FromSeconds(StartupTimeout))).ConfigureAwait(false);
+
+            if (bufferHandler.ExitCode.IsCompleted)
+            {
+                int exitCode = await bufferHandler.ExitCode.ConfigureAwait(false);
+                throw new ProcessStartupException(
```

Comment:
I don't think we need both `FunctionStartupException` and `ProcessStartupException` - they are both essentially serving the same purpose, to indicate that it wasn't possible to start the function successfully. I think we should merge them together into `FunctionStartupException`.

Reply:
Interesting point. I started off with `FunctionStartupException` and introduced the `Process` equivalent later, as it felt wrong to throw a `FunctionStartupException` when the NPM process failed.

Reply:
My question would be: what action are you going to take that's different here?

For me these feel like catastrophic exceptions, in that the entire exercise of running your tests has failed if you get either of these. So I'm not convinced there are usefully different responses to the exceptions, at which point there doesn't seem to be much point in defining different types to them.

However, I'm open to being convinced that there is a valid concrete scenario in which it's useful. Can you give me an example?

Reply:
No, just what I said in my previous comment. As you and @jongeorge1 are in agreement on this, I'll take this as direction to remove this exception :+1: